core/controllers/question_player_test.py

```diff
@@ -0,0 +1,154 @@
+# Copyright 2019 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests for the Question Player controller."""
+
+from core.domain import skill_services
+from core.tests import test_utils
+import feconf
+
+
+class SkillMasteryDataHandlerTest(test_utils.GenericTestBase):
+    """Tests update skill mastery degree."""
+
+    def setUp(self):
+        """Completes the setup for SkillMasteryDataHandler."""
+        super(SkillMasteryDataHandlerTest, self).setUp()
+        self.signup(self.NEW_USER_EMAIL, self.NEW_USER_USERNAME)
+        self.user_id = self.get_user_id_from_email(self.NEW_USER_EMAIL)
+
+        self.skill_id_1 = skill_services.get_new_skill_id()
+        self.save_new_skill(
+            self.skill_id_1, self.user_id, 'Skill Description 1')
+        self.skill_id_2 = skill_services.get_new_skill_id()
+        self.save_new_skill(
+            self.skill_id_2, self.user_id, 'Skill Description 2')
+
+    def test_put_with_valid_skill_mastery_dict(self):
+        payload = {}
+        degree_of_mastery_per_skill = {
+            self.skill_id_1: 0.3,
+            self.skill_id_2: 0.5
+        }
+        payload['degree_of_mastery_per_skill'] = degree_of_mastery_per_skill
+
+        self.login(self.NEW_USER_EMAIL)
+        csrf_token = self.get_new_csrf_token()
+        self.put_json(
+            '%s' % feconf.SKILL_MASTERY_DATA_URL,
+            payload, csrf_token=csrf_token)
+
+        self.logout()
+
+    def test_put_with_invalid_type_returns_400(self):
+        payload = {}
+        degree_of_mastery_per_skill = [self.skill_id_1, self.skill_id_2]
+        payload['degree_of_mastery_per_skill'] = degree_of_mastery_per_skill
+
+        self.login(self.NEW_USER_EMAIL)
+        csrf_token = self.get_new_csrf_token()
+        self.put_json(
+            '%s' % feconf.SKILL_MASTERY_DATA_URL,
+            payload, csrf_token=csrf_token, expected_status_int=400)
```

Comment:
In all the error cases, try to assert the error message as well. Ditto below.

Reply:
Done, except the one raises 404. I noticed when checking 404 errors, it always shows "Could not found page..." instead of the real message. Also, I checked the entire codebase and no 404 error message was asserted.